service/matching/db.go

```diff
@@ -301,6 +275,63 @@ func (db *taskQueueDB) CompleteTasksLessThan(
 	return n, err
 }
 
+// GetVersioningData returns the versioning data for this task queue. Do not mutate the returned pointer, as doing so
+// will cause cache inconsistency.
+func (db *taskQueueDB) GetVersioningData(
+	ctx context.Context,
+) (*persistencespb.VersioningData, error) {
+	db.Lock()
+	defer db.Unlock()
+
+	if db.versioningData != nil {
+		return db.versioningData, nil
+	}
+
+	tqInfo, err := db.store.GetTaskQueue(ctx, &persistence.GetTaskQueueRequest{
+		NamespaceID: db.namespaceID.String(),
+		TaskQueue:   db.taskQueueName,
+		TaskType:    db.taskType,
+	})
+	if err != nil {
+		return nil, err
+	}
+	db.versioningData = tqInfo.TaskQueueInfo.GetVersioningData()
+
+	return tqInfo.TaskQueueInfo.GetVersioningData(), nil
+}
+
+// MutateVersioningData allows callers to update versioning data for this task queue. The pointer passed to the
+// mutating function is guaranteed to be non-nil.
+func (db *taskQueueDB) MutateVersioningData(ctx context.Context, mutator func(*persistencespb.VersioningData) error) error {
+	verDat, err := db.GetVersioningData(ctx)
```

Comment:
this drops and reacquires the lock, so theoretically another update could sneak in between here and line 311 and then get lost when we write ours. to be safe I guess we should have a getVersioningDataLocked and call that within the lock here?

Reply:
Good call, fixed.